docs/kb/edk2-capsule-updates.md

```diff
@@ -0,0 +1,302 @@
+the same device (`HardwareInstance`), firmware spread out across several
+flash chips (`UpdateImageIndex`), firmware with dependencies (`Dependencies`)
+as well as for strengthening security (`MonotonicCount`):
+
+- `HardwareInstance` (defaults to `0`)
+- `UpdateImageIndex` (defaults to `1`)
+- `Dependencies` (defaults to no dependencies) which is a string in Depex format
+- `MonotonicCount` (defaults to `0`) which is a 64-bit unsigned integer that is
+  supposed to guard against replay attacks (its value is appended to the payload
+  image before signing)
+
+## Capsule creation
+
+### Preparation
+
+#### Picking a GUID
+
+Every firmware variant should have its own GUID which appears in several places:
+
+- ESRT entry describing the firmware
+- FMP instance capable of updating the firmware
+- capsule metadata
+
+Switching between firmware variables is possible by adjusting GUID and version
+fields of a capsule meant to perform the switch.
+
+#### Preparing signing keys
+
+`BaseTools/Source/Python/Pkcs7Sign` contains everything necessary for testing
+purposes and that's what will be used below.
+
+Real-world usage requires proper key management.  Key generation is covered by
+the README of `Pkcs7Sign` as well as in [tianocore's wiki][wiki-keygen].
+`Pkcs7Sign` additionally describes how to integrate generated keys into EDK2's
+build system.
+
+[wiki-keygen]: https://github.com/tianocore/tianocore.github.io/wiki/Capsule-Based-System-Firmware-Update-Generate-Keys
+
+#### Preparing `GenerateCapsule`'s input
+
+`GenerateCapsule` supports taking information via command-line keys or using a
+JSON file.  Using JSON file for configuration is more flexible and future-proof
+because options allow specifying only a single payload per capsule.  That said,
+capsule flags are accepted only in form of options.
+
+Example JSON file for a capsule that comes with an embedded FMP and uses test
+keys (relative paths assume build of Dasharo coreboot):
+
+```json
+{
+  "EmbeddedDrivers": [
+    {
+      "Driver": "../Build/DasharoPayloadPkgX64/RELEASE_COREBOOT/X64/FmpDxe.efi"
+    }
+  ],
+  "Payloads": [
+    {
+      "Payload": "../../../../../build/coreboot.rom",
+      "Guid": "00112233-4455-6677-8899-aabbccddeeff",
+      "FwVersion": "0x01020000",
+      "LowestSupportedVersion": "0x00080000",
+
+      "OpenSslSignerPrivateCertFile": "BaseTools/Source/Python/Pkcs7Sign/TestCert.pem",
+      "OpenSslOtherPublicCertFile": "BaseTools/Source/Python/Pkcs7Sign/TestSub.pub.pem",
+      "OpenSslTrustedPublicCertFile": "BaseTools/Source/Python/Pkcs7Sign/TestRoot.pub.pem"
+    }
+  ]
+}
+```
+
+Some notes on JSON:
+
+- `GenerateCapsule` warns if `EmbeddedDrivers` isn't provided, having an empty
+  array if one doesn't need to embed any drivers is enough to silence the
+  warning
+- all (even numerical) fields must be given as strings or there will be errors
+- paths to files undergo expansion of environment variables
```

Comment:
Only paths or can other fields be parametrized this way too? I'm mostly thinking about `FwVersion`, it would be tedious to change it with every release, OTOH we can `sed` this file on build time so it's not a big deal if it isn't possible.

Reply:
Only paths to files or commands use `os.path.expandvars` to process input value: https://github.com/Dasharo/edk2/blob/eb218796794a6225f4d5813de29f49d865c7f9b1/BaseTools/Source/Python/Capsule/GenerateCapsule.py#L303-L316

If depending on `jq` is fine, it can be used as a more reliable alternative to `sed`:
```
jq '.Payloads[].FwVersion = "0x10000000"' dasharo.json
```